File: cpp/067_AddBinary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        char carry = '0';
        string ret = "";
        int i = a.size()-1, j = b.size()-1;
        while( i >= 0 || j >= 0 || carry == '1'){
            char x = i >= 0 ? a[i--] : '0';
            char y = j >= 0 ? b[j--] : '0';
            ret = charAdd(x,y,carry) + ret;
        }
        return ret;
    }
    
private:
    char charAdd(char a, char b, char& carry){
        char ret = '0';
        if(a == '0'){
            if(b == '0'){
                if(carry == '0'){ //000
                    return ret;
                }else{          //001
                    ret = '1';
                    carry = '0';
                    return ret;
                }
            }else{
                if(carry == '0'){  //010
                    ret = '1';
                    return ret; 
                }else{   //011
                    return ret;
                }
            }
        }else{
            if(b == '0'){
                if(carry == '0'){ //100
                    ret = '1';
                    return ret;
                }else{ //101
                    return ret;
                }
            }else{
                if(carry == '0'){ //110
                    carry = '1';
                    return ret;
                }else{ //111
                    ret = '1';
                    return ret;
                }
            }
        }
        return ret;
    }
};

File: js/067_AddBinary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
    var carry = '0';
    var ret = "";
    var i = a.length-1, j = b.length-1;
    while(i >= 0 || j >= 0 || carry == '1'){
        var x = i >= 0 ? a[i--] : '0';
        var y = j >= 0 ? b[j--] : '0';
        var result = charAdd(x,y,carry);
        ret = result[0] + ret;
        carry = result[1];
    }
    return ret;
};

var charAdd = function(a, b, c){
    if(a === '0' && b =='0' && c == '0') return ['0', '0'];
    else if(a === '0' && b =='0' && c == '1') return ['1', '0'];
    else if(a === '0' && b =='1' && c == '0') return ['1', '0'];
    else if(a === '0' && b =='1' && c == '1') return ['0', '1'];
    else if(a === '1' && b =='0' && c == '0') return ['1', '0'];
    else if(a === '1' && b =='0' && c == '1') return ['0', '1'];
    else if(a === '1' && b =='1' && c == '0') return ['0', '1'];
    else if(a === '1' && b =='1' && c == '1') return ['1', '1'];
}